src/core/channel/compress_filter.c

```diff
@@ -0,0 +1,306 @@
+          calld->has_compression_algorithm = 1;  /* GPR_TRUE */
+          calld->compression_algorithm = GRPC_COMPRESS_NONE;
+        }
+        break;
+      case GRPC_OP_METADATA:
+        /* Parse incoming request for compression. If any, it'll be available at
+         * calld->compression_algorithm */
+        grpc_metadata_batch_filter(&(sop->data.metadata), compression_md_filter,
+                                   elem);
+        if (!calld->has_compression_algorithm) {
+          /* If no algorithm was found in the metadata and we aren't
+           * exceptionally skipping compression, fall back to the channel
+           * default */
+          calld->compression_algorithm =
+              channeld->default_compression_algorithm;
+          calld->has_compression_algorithm = 1; /* GPR_TRUE */
+        }
+        break;
+      case GRPC_OP_SLICE:
+        if (skip_compression(channeld, calld)) continue;
+        GPR_ASSERT(calld->remaining_slice_bytes > 0);
+        /* We need to copy the input because gpr_slice_buffer_add takes
+         * ownership. However, we don't own sop->data.slice, the caller does. */
+        gpr_slice_buffer_add(&calld->slices, gpr_slice_ref(sop->data.slice));
+        calld->remaining_slice_bytes -= GPR_SLICE_LENGTH(sop->data.slice);
+        if (calld->remaining_slice_bytes == 0) {
+          /* compress */
+          did_compress =
+              compress_send_sb(calld->compression_algorithm, &calld->slices);
+        }
+        break;
+      case GRPC_NO_OP:
+        ;  /* fallthrough */
+    }
+  }
+
+  /* We need to:
+   * - (OP_SLICE) If compression happened, replace the input slices with the
+   *   compressed ones.
+   * - (BEGIN_MESSAGE) Update the message info (size, flags).
+   * - (OP_METADATA) Convey the compression configuration */
+  for (i = 0, j = 0; i < send_ops->nops; ++i) {
+    grpc_stream_op *sop = &send_ops->ops[i];
+    switch (sop->type) {
+      case GRPC_OP_BEGIN_MESSAGE:
+        if (did_compress) {
+          sop->data.begin_message.length = calld->slices.length;
+          sop->data.begin_message.flags |= GRPC_WRITE_INTERNAL_COMPRESS;
+        } else {
+          /* either because the user requested the exception or because
+           * compressing would have resulted in a larger output */
+          calld->compression_algorithm = GRPC_COMPRESS_NONE;
+          /* reset the flag compression bit */
+          sop->data.begin_message.flags &= ~GRPC_WRITE_INTERNAL_COMPRESS;
+        }
+        break;
+      case GRPC_OP_METADATA:
+        grpc_metadata_batch_add_head(
+            &(sop->data.metadata), &calld->compression_algorithm_storage,
+            grpc_mdelem_ref(channeld->mdelem_compression_algorithms
+                                [did_compress ? calld->compression_algorithm
+                                              : GRPC_COMPRESS_NONE]));
+        break;
+      case GRPC_OP_SLICE:
+        if (did_compress) {
```

Comment:
Refactored and addressed (see new comments).

The "if" is also incidentally outside the loop, which may result in better branching behavior (but that wasn't the point of the refactoring, rather to make the code more readable).